shared_crates/package/src/manifest.rs

```diff
@@ -56,47 +56,102 @@ impl Manifest {
     }
 }
 
-#[derive(Deserialize, Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Hash, Default, Serialize)]
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq, Hash, Default, Serialize)]
 #[serde(transparent)]
+/// A checksummed package ID. Guaranteed to be a valid `SnakeCaseIdentifier` as well.
 pub struct PackageId(pub(crate) String);
+impl<'de> Deserialize<'de> for PackageId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        PackageId::new(&String::deserialize(deserializer)?).map_err(serde::de::Error::custom)
+    }
+}
 impl PackageId {
+    const DATA_LENGTH: usize = 12;
+    const CHECKSUM_LENGTH: usize = 8;
+    const TOTAL_LENGTH: usize = Self::DATA_LENGTH + Self::CHECKSUM_LENGTH;
+
     pub fn as_str(&self) -> &str {
         &self.0
     }
 
+    /// Attempts to create a new package ID from a string.
+    pub fn new(id: &str) -> Result<Self, String> {
+        Self::validate(id)?;
+        Ok(Self(id.to_string()))
+    }
+
     /// Generates a new package ID.
-    // TODO: suffix with checksum
     pub fn generate() -> Self {
-        const NUMERALS: [u8; 10] = [b'0', b'1', b'2', b'3', b'4', b'5', b'6', b'7', b'8', b'9'];
-        const ALPHABET: [u8; 26] = [
-            b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', b'i', b'j', b'k', b'l', b'm', b'n',
-            b'o', b'p', b'q', b'r', b's', b't', b'u', b'v', b'w', b'x', b'y', b'z',
-        ];
-
-        let mut rng = rand::thread_rng();
-        let mut output = vec![0u8; 32];
-        output[0] = *ALPHABET.choose(&mut rng).unwrap();
-        for i in 1..output.len() {
-            let idx = rng.gen_range(0..(NUMERALS.len() + ALPHABET.len()));
-            output[i] = if idx < NUMERALS.len() {
-                NUMERALS[idx]
-            } else {
-                ALPHABET[idx - NUMERALS.len()]
-            };
+        // TODO: see if there's a more intelligent way to ensure the first character
+        // is always alphabetic
```

Comment:
Almost like BTC mining 😝

Reply:
The comparison was not lost on me and it makes me sad 😂